File: 백준/Gold/21610. 마법사 상어와 비바라기/마법사 상어와 비바라기.java
import com.sun.security.jgss.GSSUtil;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.lang.reflect.Array;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String str = br.readLine();
        StringTokenizer st = new StringTokenizer(str," ");
         int N = Integer.parseInt(st.nextToken());
         int M = Integer.parseInt(st.nextToken());
         int [][] board = new int[N][N];
         int [][] move = new int[M][2];
         boolean[][] visit = new boolean[N][N];
        for (int i = 0; i < N; i++){
            st = new StringTokenizer(br.readLine()," ");
            for (int j = 0; j < N; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        for (int i = 0; i < M; i++){
            st = new StringTokenizer(br.readLine()," ");
            for (int j = 0; j < 2; j++){
                move[i][j] = Integer.parseInt(st.nextToken());
            }
        }





        ArrayList<int[]> clouds = new ArrayList<>();
        int[] dx = {0,0,-1,-1,-1,0,1,1,1};
        int[] dy = {0,-1,-1,0,1,1,1,0,-1};
        int[][] init = {{N-1,0},{N-1,1},{N-2,0},{N-2,1}};
        // 비바라기 시전
        for(int[] e : init) {
            clouds.add(e);
        }
        // M번의 이동 시작
        for (int i = 0; i < M; i++) {
            int[] currMove = move[i];
            // 현재 명령의 방향과 속도 설정
            int d = currMove[0];
            int s = currMove[1];
            // 구름 이동 -> 모든 구름이 d방향으로 s칸 만큼 이동 + 비내리기
            for (int j = 0; j < clouds.size(); j++) {
                int[] arr = clouds.get(j);
                int nx = (arr[0] + dx[d] * (s%N));
                int ny = (arr[1] + dy[d] * (s%N));
                if (nx >= 0) nx = nx % N;
                else if (nx < 0) nx = N + nx;
                if (ny >= 0) ny = ny % N;
                else if (ny < 0) ny = N + ny;
                board[nx][ny] += 1;
                int[] temp = {nx,ny}; // 구름이 최종 이동한 위치
                clouds.set(j,temp);
                visit[nx][ny] = true;

            }
            ArrayList<int[]> newClouds = new ArrayList<>();
            ArrayList<String> newRemember = new ArrayList<>();
            // 물복사버그
            for (int j = 0; j < clouds.size(); j++) {
                int[] arr = {2,4,6,8};
                int x = clouds.get(j)[0];
                int y = clouds.get(j)[1];
                int cnt = 0;
                for (int dir : arr) { // 대각선 4방향 체크
                    int nx = x + dx[dir];
                    int ny = y + dy[dir];
                    if (nx < 0 || nx >= N || ny < 0 || ny >= N) continue;
                    if (board[nx][ny] >= 1) cnt ++;
                }
                board[x][y] += cnt;
            }

            // 물의 양이 2 이상인 칸에 구름 생성, 물 양 - 2 이때, 원래 구름이 있던 곳은 안돼!
            for (int x = 0; x < N; x++) {
                for (int y = 0; y < N; y++) {
                    int[] arr = {x,y};
                    if (board[x][y] >= 2 && !visit[x][y]) {
                        board[x][y] -= 2;
                        newClouds.add(arr);
                        newRemember.add(Arrays.toString(arr));
                    }

                }
            }

            clouds = newClouds;
            visit = new boolean[N][N];


        }
        // M번 이동 종료 후 물의 양 합 구하기
        int ans = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                ans += board[i][j];
            }
        }

        System.out.println(ans);
    }
}

File: 백준/Gold/21610. 마법사 상어와 비바라기/마법사 상어와 비바라기.py
import sys
input = sys.stdin.readline


# N*N 보드 처음과 끝이 연결되어있음
# 최초 비바라기 시전 시 좌측 하단 4칸짜리 비바라기 생성
# 이동 명령
    # 구름이 방향에 따라 이동 (처음과 끝 연결)
    # 구름이 생기는 좌표를 list로 담기
    # 구름list를 돌면서 board[x][y] 증가
    # 구름list의 (x,y) 좌표 돌면서 추가 물 계산 (이때는 사이클 x)
    # if board[x][y] >= 2 and not in 구름list :
        # board[x][y] -= 2
        # newclouds.append((x,y))

# 입력
dx = (0,0,-1,-1,-1,0,1,1,1)
dy = (0,-1,-1,0,1,1,1,0,-1)
N,M = map(int,input().split())
board = [list(map(int,input().split())) for _ in range(N)]
query = [list(map(int,input().split())) for _ in range(M)]

def move(clouds,d,s):
    newClouds = []
    for e in clouds :
        x,y = e
        nx,ny = (x + dx[d] * s) % N, (y + dy[d] * s) % N
        nx = (N + nx) if nx < 0 else nx
        ny = (N + ny) if ny < 0 else ny
        newClouds.append((nx,ny))
    return newClouds
# 최초 구름 init
clouds = []
for e in [(N-1,0),(N-1,1),(N-2,0),(N-2,1)]:
    clouds.append(e)
# 구름 이동 명령 시작
for q in query :
    d,s = q
    # 구름이 d방향으로 s만큼 이동
    clouds = move(clouds,d,s)
    # 구름에서 비 내려서 구름 칸 증가
    for e in clouds :
        x,y = e
        board[x][y] += 1
    # 구름이 있던 칸 물복사 버그 시전
    for e in clouds :
        x,y = e
        cnt = 0
        # 대각선만 체크
        for i in [2,4,6,8]:
            nx,ny = x + dx[i], y + dy[i]
            # 사이클 형성 x -> 범위 제대로 잡아줘야함
            if 0 <= nx < N and 0 <= ny < N:
                if board[nx][ny] > 0 : cnt += 1
        board[x][y] += cnt
    # board를 돌면서 새로운 구름 생성
    setClouds = set(clouds) # in 메서드를 빠르게 돌리기 위한 set 사용
    newClouds = []
    for i in range(N):
        for j in range(N):
            if board[i][j] >= 2 and (i,j) not in setClouds:
                newClouds.append((i,j))
                board[i][j] -= 2
    clouds = newClouds

ans = 0
for e in board :
    ans += sum(e)
print(ans)


